### PR TITLE
planner: make unsafe substitute can be used in generated column

### PR DESCRIPTION
**Pull request by xiongjiwei**
Signed-off-by: xiongjiwei <xiongjiwei1996@outlook.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/35490

Problem Summary:

### What is changed and how it works?

add a system variable `tidb_enable_unsafe_substitute`, it controls if a string type generate column can be substituted by an expression even if the expression's length is not the same as the column definition.
use this variable has a chance to query the wrong result, the user should guarantee they do not break the table column constraint. 
for more detail, see https://github.com/pingcap/tidb/issues/35490#issuecomment-1211658886

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
add a system variable `tidb_enable_unsafe_substitute` to control if an expression can be substituted by a column unsafely
```

### CODE REVIEW

File: parser/types/field_type.go [hunk 1]
```diff
@@ -273,6 +273,24 @@ func (ft *FieldType) Equal(other *FieldType) bool {
 	return true
 }
 
+// UnsafeEqual checks whether two string FieldType objects are equal without flen.
+func (ft *FieldType) UnsafeEqual(other *FieldType, unsafe bool) bool {
```
Comment (wjhuang2016) [inline parser/types/field_type.go:277]: Not a good name...

File: cmd/explaintest/r/explain_generate_column_substitute.result [hunk 2]
```diff
@@ -642,3 +642,87 @@ HashJoin	9.99	root		inner join, equal:[eq(test.t2.c_decimal, test.t1.c_decimal)]
...
+drop table person;
+create table t(a char(5), b char(10) as (concat(a, a)), index bx(b));
+insert into t(a) values ('aaaaa');
+select * from t;
+a	b
+aaaaa	aaaaaaaaaa
+select * from t ignore index(bx) where concat(a, a) = 'aaaaaaaaaa';
+a	b
+aaaaa	aaaaaaaaaa
+desc select * from t ignore index(bx) where concat(a, a) = 'aaaaaaaaaa';
```
Comment (xuyifangreeneyes) [inline cmd/explaintest/r/explain_generate_column_substitute.result:714]:
```suggestion
desc format = 'brief' select * from t ignore index(bx) where concat(a, a) = 'aaaaaaaaaa';
```
`format = 'brief'` can remove the suffix number for operators, which avoids unnecessary explain result changes when modifying optimizer code.

Comment (xiongjiwei) [reply cmd/explaintest/r/explain_generate_column_substitute.result]: done

File: parser/types/field_type.go (hunk 1 above)
Comment (xiongjiwei) [reply parser/types/field_type.go]: renamed

File: planner/core/rule_generate_column_substitute.go [hunk 3]
```diff
@@ -66,7 +66,7 @@ func collectGenerateColumn(lp LogicalPlan, exprToColumn ExprColumnMap) {
 			if colInfo.IsGenerated() && !colInfo.GeneratedStored {
 				s := ds.schema.Columns
 				col := expression.ColInfo2Col(s, colInfo)
-				if col != nil && col.GetType().Equal(col.VirtualExpr.GetType()) {
+				if col != nil && col.GetType().PartialEqual(col.VirtualExpr.GetType(), lp.SCtx().GetSessionVars().EnableUnsafeSubstitute) {
```
Comment (xuyifangreeneyes) [inline planner/core/rule_generate_column_substitute.go:69]: Maybe we can add an example to explain why we want to enable unsafe substitution to make it easy to understand.

**Review by xuyifangreeneyes — APPROVED** [review]
rest LGTM

File: planner/core/rule_generate_column_substitute.go (hunk 3 above)
Comment (xiongjiwei) [reply planner/core/rule_generate_column_substitute.go:69]: done, add an issue link

**Review by wjhuang2016 — APPROVED** [review]

**Review by bb7133 — APPROVED** [review]
LGTM

**Review by bb7133 — APPROVED** [review]
LGTM

**Pull request merged**